ShaderTweaks Phase 1 (Part 1)

johnhaddon opened the pull request:
This switches Gaffer's internal representation for assigned shaders and lights, adopting the ShaderNetwork class introduced in https://github.com/ImageEngine/cortex/pull/890. This PR adds no functionality whatsoever, just updating everything so that what we had before works with the new data representation.

I'll continue to work on this, first implementing the per-component connections requested in https://github.com/GafferHQ/gaffer/issues/2839, and then introducing the ShaderTweaks node for https://github.com/GafferHQ/gaffer/issues/2685. But I wanted to get a PR up now so that a) I can get some early feedback and b) people with affected code can adapt it and report back any problems.

Note that the tests won't pass just yet, due to the dependency on https://github.com/ImageEngine/cortex/pull/890.

johnhaddon commented on python/GafferArnoldTest/IECoreArnoldPreviewTest/RendererTest.py:
```diff
@@ -259,39 +271,37 @@ def testShaderNames( self ) :
 			self.assertEqual( len( [ i for i in shaderNames if i.split(":")[-1] == "myHandle" ] ), 2 )
 
-	def testShaderNodeConnectionType( self ) :
+	def testShaderConnections( self ) :
```
@danieldresser-ie, this could do with your eyes. I've intentionally removed the support for connections to non-shader nodes that you introduced back in https://github.com/GafferHQ/gaffer/commit/584671517170422a2102788c00548fd79f1cf440#diff-7e15ae85c10bcfa25e451ada1ca4e068. I did this partly on the assumption that it's no longer needed (I may be wrong, but I think it was for early light filter support, and we've got a more official mechanism for that now), partly on the basis that it really doesn't make much sense, and partly on the basis that it allows further code cleanup. Can you think of any reason it's not OK?

On the plus side, I have been able to revive the commented test for connections to array shader parameters, because Arnold now includes ramp_rgb in the standard distribution.

johnhaddon commented:
I've just force-pushed an update. It's largely the same as before except with tweaks required by the final form that ShaderNetwork took, and with various fixes I've made since. It also contains a final series of commits that add support for component-level connections for colour and vector parameters. I've uploaded the updated dependencies for Linux, and Mac is on the way, so in theory the tests should now pass.

UI components to support the component-level connections to follow. I think this is worth reviewing/merging on its own though, particularly as we have some internal code at IE that will need to be adapted to the new ShaderNetwork representation.

johnhaddon commented:
So, I discovered a last minute problem with a recent change I made in Cortex, which is the reason the tests are failing here. I have a [Cortex PR](https://github.com/ImageEngine/cortex/pull/896) up to fix that, and once it's merged I'll build new dependencies packages.

johnhaddon commented:
We really need to prioritise this one for review - see comments on #2913.

donboie commented on src/GafferScene/Shader.cpp:
```diff
@@ -383,28 +362,25 @@ class Shader::NetworkBuilder
-				{
-					values[parameterName] = value;
-				}
+				this->addParameter( parameter, parameterName, shader, connections );
```
`this->` can be dropped

johnhaddon replied:
Done in aca2ce1.

johnhaddon commented on python/GafferArnoldTest/ArnoldRenderTest.py:
```diff
@@ -982,6 +984,59 @@ def testAbortRaises( self ) :
+
+		swizzle = GafferOSL.OSLShader()
+		swizzle.loadShader( "MaterialX/mx_swizzle_color_float" )
+		swizzle["parameters"]["in"].setValue( imath.Color3f( 1 ) )
```
Good idea - addressed in 35101c5. I added the test to expose a bug that I introduced during the refactor, and which resulted in a total failure to shade at all, hence the exact colour didn't matter. But your suggestion makes it more likely it'll catch other problems in the future.

johnhaddon commented on python/GafferArnoldTest/ArnoldRenderTest.py:
```diff
@@ -982,6 +984,59 @@ def testAbortRaises( self ) :
 
 		self.assertRaisesRegexp( RuntimeError, "Render aborted", s["render"]["task"].execute )
 
+	def testMXShaders( self ) :
```
Renamed to `testOSLShaders` in 35101c5, to clarify that we're not really testing MaterialX at all, but rather our Arnold/OSL integration.

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/ShaderNetworkAlgo.cpp:
```diff
@@ -0,0 +1,243 @@
+//////////////////////////////////////////////////////////////////////////
```
Fair point. For the record, this file contains the most substantial changes of all in this PR - it's _much_ simpler now. Fortunately, it's changed enough that I think reviewing it as new code rather than as diffs is probably the most effective approach anyway.

donboie commented on src/GafferScene/Shader.cpp:
```diff
@@ -249,124 +253,99 @@ class Shader::NetworkBuilder
+			for( const auto &c : inputConnections )
 			{
-				result += "." + outputParameter->relativeName( shaderNode->outPlug() );
+				m_network->addConnection( { c.source, { handleAndHash.handle, c.destination.name } } );
```
Just for the record, I disagree. Having a bunch of nested  ```{ { }, {} } ``` makes things more confusing at the call site to me.

danieldresser-ie replied:
Also for the record, I'm with John on this one.  I like special purpose syntax that favors conciseness over generality ( though it's pretty rare that John agrees with me on this ... )

johnhaddon commented on src/GafferScene/Shader.cpp:
```diff
@@ -249,124 +253,99 @@ class Shader::NetworkBuilder
 		}
 
-		IECoreScene::Shader *shader( const Shader *shaderNode )
+		IECore::InternedString handle( const Shader *shaderNode )
```
I don't disagree, but it's equivalent to the previous name, and I'm blanking on a better name right now.

johnhaddon commented:
Thanks for the review Don. I've addressed some of your comments and merged. My feeling on the others is that they're more a matter of philosophy/taste - happy to continue the discussion separately, but I don't think they merit blocking progress (and I need to merge this before I can merge some of Daniel's stuff).
